### PR TITLE
nGraph Integration with TensorFlow 

### PR DESCRIPTION
**Pull request by avijit-nervana**
This is the first revision of the integration of nGraph with TensorFlow. The nGraph provides a JIT compilation and execution of TensorFlow computation graphs. To enable nGraph, select the configuration option during running configure.

### CODE REVIEW

**Comment by michaelisard** [comment]
I don't think I'm a good reviewer for this since it seems to be almost entirely build-system edits and I'm not very familiar with that.

**Comment by mrry** [comment]
Nor am I particularly. @avijit-nervana Have you been in touch with anyone on the TensorFlow team who might have more context for this PR?

**Comment by avijit-nervana** [comment]
@mrry Thanks for the follow up. Yes - we have been in discussiosn with Rajat Monga and Tatiana Shpeisman about this integration. Tatiana mentioned that she will be  following up on this PR.

**Comment by caisq** [comment]
cc @tatianashp

**Comment by tatianashp** [comment]
I'll review and pull in others as necessary.

File: third_party/ngraph/ngraph_tf.BUILD [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+    srcs = [
+        "lib/libiomp5.so",
+        "lib/libmklml_intel.so",
+    ],
+    visibility = ["//visibility:public"],
+)
+
+cc_library(
+    name = "ngraph_tf",
+    srcs =
```
Comment (aaroey) [inline third_party/ngraph/ngraph_tf.BUILD]: nit: please put the open bracket on the this line. It'll be useful to run [buildifier](https://github.com/bazelbuild/buildtools/tree/master/buildifier) on all the BUILD files.

**Review by aaroey — CHANGES_REQUESTED** [review]
Thanks for the PR, here are some initial comments. I think someone will need to set up internal (dummy) build environment for ngraph before this PR can be built or can pass the tests.

File: third_party/ngraph/build_defs.bzl [hunk 2]
```diff
@@ -0,0 +1,16 @@
+def clean_dep(dep):
+  return str(Label(dep))
+
+def if_ngraph(a):
```
[the remark this answers is not in the document]
Comment (avijit-nervana) [reply third_party/ngraph/build_defs.bzl]: Ok - changed.

**Comment by avijit-nervana** [comment]
@aaroey Thanks for the comments. I have updated the PR. There is no need to setup any specific build environment for building nGraph on Ubuntu 16.04 and macOS. Just selecting nGraph build option during the configure step will suffice to built TensorFlow with nGraph. Please let us know if you have additional questions or comments.

**Comment by aaroey** [comment]
Thanks for the fix @avijit-nervana. Would you please help to fix the conflicts?
Also +@gunan.

**Comment by avijit-nervana** [comment]
@aaroey Resolved the conflicts. Please let me know if there are further changes needed. Thanks!

File: tensorflow/BUILD [hunk 3]
```diff
@@ -432,6 +441,8 @@ load(
     "if_mkl",
 )
 
+load("//third_party/ngraph:build_defs.bzl", "if_ngraph")
```
Comment (aaroey) [inline tensorflow/BUILD]: Please put this at the top of the file.

File: WORKSPACE [hunk 4]
```diff
@@ -78,4 +78,4 @@ new_http_archive(
         "http://storage.googleapis.com/download.tensorflow.org/models/speech_commands_v0.01.zip",
         "http://download.tensorflow.org/models/speech_commands_v0.01.zip",
     ],
-)
+)
```
Comment (aaroey) [inline WORKSPACE]: Revert?

File: tensorflow/BUILD [hunk 5]
```diff
@@ -387,6 +387,7 @@ config_setting(
     define_values = {
         "dynamic_loaded_kernels": "true",
     },
+    visibility = ["//visibility:public"],
```
Comment (aaroey) [inline tensorflow/BUILD]: Just curious, why do you need this to be public?

File: tensorflow/core/BUILD [hunk 6]
```diff
@@ -2521,7 +2523,7 @@ tf_cuda_library(
             "//third_party/mkl:intel_binary_blob",
             "@mkl_dnn",
         ],
-    ),
+    ),
```
Comment (aaroey) [inline tensorflow/core/BUILD]: Please revert.

File: tensorflow/core/platform/default/build_config.bzl [hunk 7]
```diff
@@ -708,3 +708,4 @@ def tf_additional_binary_deps():
           "//third_party/mkl:intel_binary_blob",
       ],
   )
+
```
Comment (aaroey) [inline tensorflow/core/platform/default/build_config.bzl]: Please revert.

File: tensorflow/python/BUILD [hunk 8]
```diff
@@ -3808,7 +3809,8 @@ tf_py_wrap_cc(
          tf_additional_plugin_deps() +
          tf_additional_verbs_deps() +
          tf_additional_mpi_deps() +
-         tf_additional_gdr_deps()),
+         tf_additional_gdr_deps())+
```
Comment (aaroey) [inline tensorflow/python/BUILD]: nit: please add a space before `+`.

File: tensorflow/tensorflow.bzl [hunk 9]
```diff
@@ -29,6 +29,11 @@ load(
     "//third_party/mkl_dnn:build_defs.bzl",
     "if_mkl_open_source_only",
 )
+load(
```
Comment (aaroey) [inline tensorflow/tensorflow.bzl]: nit: please add a empty line before `load(`

File: tensorflow/workspace.bzl [hunk 10]
```diff
@@ -825,6 +824,39 @@ def tf_workspace(path_prefix = "", tf_repo_name = ""):
         strip_prefix = "rules_android-0.1.1",
     )
 
+    tf_http_archive(
```
Comment (aaroey) [inline tensorflow/workspace.bzl]: Please revert the change at line 9.

Comment (aaroey) [inline tensorflow/workspace.bzl]: +@gunan for comments on this dependencies.

File: third_party/ngraph/LICENSE [hunk 11]
```diff
@@ -0,0 +1,201 @@
+                                 Apache License
```
Comment (aaroey) [inline third_party/ngraph/LICENSE]: +@gunan, @tfboyd for license.

**Review by aaroey — CHANGES_REQUESTED** [review]

File: third_party/ngraph/ngraph_tf.BUILD [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+cc_library(
+    name = "ngraph_libs_linux",
+    srcs = [
+        "lib/libiomp5.so",
+        "lib/libmklml_intel.so",
+    ],
+    visibility = ["//visibility:public"],
+)
+
+cc_library(
```
Comment (aaroey) [inline third_party/ngraph/ngraph_tf.BUILD]: Thanks for the fixes. There is one more request from our build team: could you make this a `cc_binary` with `linkshared=1` and test if it works? This aligns with TF's roadmap about tf.contrib and will help to make future integration/migration easier.

Other than that, this PR looks good to me.

Comment (avijit-nervana) [reply third_party/ngraph/ngraph_tf.BUILD]: Thanks @aaroey for the feedback. I am going to try this an report back.

Comment (avijit-nervana) [reply third_party/ngraph/ngraph_tf.BUILD]: @aaroey I was able to use `linkshared=1` a shared library for the unit tests (i.e., `ngraph_tf_tests`) but for the TF python module (i.e., `pywrap_tensorflow_internal`) I don't know how to link it. If you can provide some help that I can try that as well.

File: third_party/ngraph/LICENSE (hunk 11 above)
Comment (thirupalanisamy) [reply third_party/ngraph/LICENSE]: Thanks @tfboyd !

We further reviewed the licenses of the dependency libraries  (ngraph, nlohmann_json_lib, ngraph_tf ). They are notice-only licenses and already approved.

Comment (aaroey) [reply third_party/ngraph/LICENSE]: Thanks @thirupalanisamy and @tfboyd !

File: third_party/ngraph/ngraph_tf.BUILD (hunk 12 above)
Comment (aaroey) [reply third_party/ngraph/ngraph_tf.BUILD]: @avijit-nervana Do you mean there were errors when linking it with `pywrap_tensorflow_internal`, or it can dynamically linked but not be able to be loaded when user `import tensorflow`?

Comment (avijit-nervana) [reply third_party/ngraph/ngraph_tf.BUILD]: @aaroey I am seeing `bazel` errors. I was able to add the `@ngraph_tf//:libngraph_tf.so` to the `//tensorflow:libtensorflow_cc.so` target by adding the following line:
```
srcs = if_ngraph(["@ngraph_tf//:libngraph_tf.so"]),
```
However, don't know how to add this dependency to the `pywrap_tensorflow_internal` as this is a `tf_py_wrap_cc(...)` and looks like only soiurce it takes is a swig input file.

What's the recommended way to add the .so to the python wrapper? 

Also - if an external application uses the TensorFlow C++ API (e.g., a TensorFlow serving container) - then does the external application also need to link with `@ngraph_tf//:libngraph_tf.so`?

Comment (gunan) [reply third_party/ngraph/ngraph_tf.BUILD]: So what we had in mind is, rather than explicit dynamic linking in tensorflow, ngraph build and users use this mechanism.

1 - When building, core TF only needs to expose the registration mechanism for optimizers.
2 - We build TF without ngraph
3 - You build ngraph to your specifications.
4 - Users load TF.
5 - Users dlopen ngraph library. This triggers a static initializer that registers ngraph optimizer with the core TF optimizer registry.
6 - Users then use ngraph code.

So following the above model:
- TF should not explicitly link against ngraph
- ngraph should explicitly link against libtensorflow_framework.so
- no explicit bazel dependency should be needed
- external applications can choose to explicitly link against ngraph, or they can simply dlopen ngraph. Either should just work.

Comment (aaroey) [reply third_party/ngraph/ngraph_tf.BUILD]: Thanks @avijit-nervana for the explanation and @gunan for the suggestions. I also tested the approach about `cc_binary` with `linkshared=1`, and it turns out it's not working since `cc_binary` is not expected in `tf_py_wrap_cc`, and we'll need to implement what @gunan says when the mechanism is ready.

At this moment, since the mechanism mentioned is not fully implemented/tested yet, I think we can keep the current approach and do the migration later which will make things easier and less error-prone.

**Review by aaroey — DISMISSED** [review]

File: third_party/ngraph/ngraph_tf.BUILD (hunk 12 above)
Comment (avijit-nervana) [reply third_party/ngraph/ngraph_tf.BUILD]: @aaroey Thanks for the approval. Since there is now a conflict with master - do you want me to resolve this?

Comment (aaroey) [reply third_party/ngraph/ngraph_tf.BUILD]: @avijit-nervana yes, please help to resolve it.

Comment (avijit-nervana) [reply third_party/ngraph/ngraph_tf.BUILD]: @aaroey Done.

**Review by aaroey — APPROVED** [review]
